x/multisig/decorator.go

```diff
@@ -44,36 +42,52 @@ func (d Decorator) Deliver(ctx weave.Context, store weave.KVStore, tx weave.Tx,
 
 func (d Decorator) withMultisig(ctx weave.Context, store weave.KVStore, tx weave.Tx) (weave.Context, error) {
 	if multisigContract, ok := tx.(MultiSigTx); ok {
-		// does tx have multisig ?
-		id := multisigContract.GetMultisig()
-		if id == nil {
-			return ctx, nil
-		}
+		ids := multisigContract.GetMultisig()
+		for _, contractID := range ids {
+			if contractID == nil {
+				return ctx, nil
+			}
 
-		// load contract
-		obj, err := d.bucket.Get(store, id)
-		if err != nil {
-			return ctx, err
-		}
-		if obj == nil || (obj != nil && obj.Value() == nil) {
-			return ctx, ErrContractNotFound(id)
-		}
-		contract := obj.Value().(*Contract)
+			// check if we already have it
+			if d.auth.HasAddress(ctx, MultiSigCondition(contractID).Address()) {
+				return ctx, nil
+			}
 
-		// retrieve sigs
-		var sigs []weave.Address
-		for _, sig := range contract.Sigs {
-			sigs = append(sigs, sig)
-		}
+			// load contract
+			contract, err := d.getContract(store, contractID)
+			if err != nil {
+				return ctx, err
+			}
 
-		// check sigs
-		authenticated := x.HasNAddresses(ctx, d.auth, sigs, int(contract.ActivationThreshold))
-		if !authenticated {
-			return ctx, ErrUnauthorizedMultiSig(id)
-		}
+			// collect all sigs
+			sigs := make([]weave.Address, len(contract.Sigs))
```

Comment:
It might be nice to make that a method on the contract, so
`sigs := contract.GetSigsAddresses()` works and no need to do this remapping in every usage, just once.